core/common/src/main/java/alluxio/underfs/ObjectMultiRangeInputStream.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for reading data using range queries.
+ */
+@NotThreadSafe
+public abstract class ObjectMultiRangeInputStream extends InputStream {
+
+  /** Has the stream been closed. */
+  private boolean mClosed;
+  /** The backing input stream. */
+  protected InputStream mStream;
+  /** The current position of the stream. */
+  protected long mPos;
+  /** Position the current stream was open till (inclusive). */
+  protected long mEndPos;
+
+  @Override
+  public void close() throws IOException {
+    if (!mClosed) {
+      closeStream();
+    }
+    mClosed = true;
+  }
+
+  @Override
+  public int read() throws IOException {
+    if (mStream == null) {
+      openStream();
+    }
+    int value = mStream.read();
+    if (value != -1) { // valid data read
+      mPos++;
+      closeStreamIfBoundary();
+    }
+    return value;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int offset, int length) throws IOException {
+    if (length == 0) {
+      return 0;
+    }
+    if (mStream == null) {
+      openStream();
+    }
+    int read = mStream.read(b, offset, length);
+    if (read != -1) {
+      mPos += read;
+      closeStreamIfBoundary();
+    }
+    return read;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    if (n <= 0) {
+      return 0;
+    }
+    closeStream();
+    mPos += n;
+    openStream();
+    return n;
+  }
+
+  /**
+   * Close the current stream if the boundary for a range is crossed.
+   */
+  private void closeStreamIfBoundary() throws IOException {
+    if (mPos > mEndPos) {
```

Comment:
If `mPos == mEndPos`, should the stream also be closed? Otherwise the next read can possibly return `-1` prematurely?